### PR TITLE
Introduce listener mechanism in InternalSnapshotsInfoService to help testing

### PR DESCRIPTION
**Pull request by tlrx**
This pull request introduces a change in `InternalSnapshotsInfoService` that allows listeners to register themselves to be notified when fetching the size of a snapshotted shard is a success or a failure.

The motivation is to help and to simplify tests like `CcrRepositoryIT` which are verifying that fetching the size from the remote leader cluster is successful (or not). Because the sizes of the snapshot shards are only kept in a very short period of time the test cannot rely on analyzing the cluster state and querying the `SnapshotsInfoService`, and this can trigger failures like #64167.

Closes #64167

### CODE REVIEW

File: x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java [hunk 1]
```diff
@@ -513,99 +511,65 @@ public void testCcrRepositoryFetchesSnapshotShardSizeFromIndexShardStoreStats()
...
+            .indices(leaderIndex).indicesOptions(indicesOptions).renamePattern("^(.*)$")
+            .renameReplacement(followerIndex)
+            .masterNodeTimeout(TimeValue.MAX_VALUE)
+            .indexSettings(Settings.builder()
+                .put(IndexMetadata.SETTING_INDEX_PROVIDED_NAME, followerIndex)
+                .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true));
 
-            final SnapshotsInfoService snapshotsInfoService = getFollowerCluster().getCurrentMasterNodeInstance(SnapshotsInfoService.class);
-            assertBusy(() -> {
-                SnapshotShardSizeInfo snapshotShardSizeInfo = snapshotsInfoService.snapshotShardSizes();
```
Comment (original-brownbear) [inline x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java:564]: Couldn't we run this busy assert on the CS thread as soon as we observe the restore/recovery there? That way the recovery is not removed concurrently and we get a consistent view of the `InternalSnapshotsInfoService` without having to complicate production code?

**Review by original-brownbear — COMMENTED** [review]
Thanks for tracking this down Tanguy!

One question/suggestion about an alternative solution that might be simpler :)

File: x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java:564]: I should have thought about it myself, it seems obvious, there's not need to complexify production like I did. Thanks Armin!

**Comment by tlrx** [comment]
Closed in favor of #64976

**Pull request closed without merging**